example/gmres/gmres.hpp

```diff
@@ -0,0 +1,307 @@
+#include<KokkosBlas.hpp>
+#include<KokkosBlas3_trsm.hpp>
+#include<KokkosSparse_spmv.hpp>
+
+// This struct is returned to the user to give solver
+// statistics and convergence status. 
+struct GmresStats {
+  int numIters;
+  double endRelRes;
+  enum FLAG { Conv, NoConv, LOA };
+  FLAG convFlagVal;
+  std::string convFlag() {
+    switch(convFlagVal){
+      case Conv:
+        return "Converged";
+      case NoConv:
+        return "Not Converged";
+      case LOA:
+        return "Solver has had loss of accuracy.";
+      default:
+        return "Flag not defined.";
+    }
+  }
+};
+
+template< class ScalarType, class Layout, class EXSP, class OrdinalType = int > 
+  GmresStats gmres( const KokkosSparse::CrsMatrix<ScalarType, OrdinalType, EXSP> &A, 
+                    const Kokkos::View<ScalarType*, Layout, EXSP> &B,
+                    Kokkos::View<ScalarType*, Layout, EXSP> &X, 
+                    const typename Kokkos::Details::ArithTraits<ScalarType>::mag_type tol = 1e-8, 
+                    const int m=50, 
+                    const int maxRestart=50, 
+                    const std::string ortho = "CGS2"){
+
+  typedef Kokkos::Details::ArithTraits<ScalarType> AT;
+  typedef typename AT::val_type ST; // So this code will run with ScalarType = std::complex<T>.
+  typedef typename AT::mag_type MT; 
+  ST one = AT::one();
+  ST zero = AT::zero();
+
+  typedef Kokkos::View<ST*, Layout, EXSP> ViewVectorType;
+  typedef Kokkos::View<ST*, Kokkos::LayoutRight, Kokkos::HostSpace> ViewHostVectorType; 
+  typedef Kokkos::View<ST**, Layout, EXSP> ViewMatrixType;
+
+  unsigned int n = A.numRows();
+
+  // Check compatibility of dimensions at run time.
+  if ( n != unsigned(A.numCols()) ){
+    std::ostringstream os;
+    os << "gmres: A must be a square matrix: " 
+      << "numRows: " << n << "  numCols: " << A.numCols();
+      Kokkos::Impl::throw_runtime_exception (os.str ());
+  }
+
+  if (X.extent(0) != B.extent(0) ||
+      X.extent(0) != n ) {
+    std::ostringstream os;
+    os << "gmres: Dimensions of A, X, and B do not match: "
+       << "A: " << n << " x " << n << ", X: " << X.extent(0) 
+       << "x 1, B: " << B.extent(0) << " x 1";
+    Kokkos::Impl::throw_runtime_exception (os.str ());
+  }
+  //Check parameter validity:
+  if(m <= 0){
+    throw std::invalid_argument("gmres: please choose restart size m greater than zero.");
+  }
+  if(maxRestart <= 0){
+    throw std::invalid_argument("gmres: Please choose maxRestart greater than zero.");
+  }
+
+  bool converged = false;
+  int cycle = 0; // How many times have we restarted? 
+  int numIters = 0;  //Number of iterations within the cycle before convergence.
+  MT nrmB, trueRes, relRes, shortRelRes;
+  GmresStats myStats;
+  
+  std::cout << "Convergence tolerance is: " << tol << std::endl;
+
+  ViewVectorType Xiter("Xiter",n); //Intermediate solution at iterations before restart. 
+  ViewVectorType Res(Kokkos::ViewAllocateWithoutInitializing("Res"),n); //Residual vector
```

Comment:
See previous comment about the deprecation of `Kokkos::ViewAllocateWithoutInitializing`